src/Test/NodeTestCase.php

```diff
@@ -51,24 +65,53 @@ protected function getCompiler(?Environment $environment = null)
         return new Compiler($environment ?? $this->getEnvironment());
     }
 
+    /**
+     * @final since Twig 3.13
+     */
     protected function getEnvironment()
     {
-        if (!$this->currentEnv) {
-            $this->currentEnv = new Environment(new ArrayLoader());
-        }
+        return $this->currentEnv ??= static::createEnvironment();
+    }
 
-        return $this->currentEnv;
+    protected static function createEnvironment(): Environment
+    {
+        return new Environment(new ArrayLoader());
     }
 
+    /**
+     * @deprecated since Twig 3.13, use createVariableGetter() instead.
+     */
     protected function getVariableGetter($name, $line = false)
+    {
+        trigger_deprecation('twig/twig', '3.13', 'Method "%s()" is deprecated, use "createVariableGetter()" instead.', __METHOD__);
+
+        return self::createVariableGetter($name, $line);
+    }
+
+    final protected static function createVariableGetter($name, $line = false): string
     {
         $line = $line > 0 ? "// line $line\n" : '';
 
         return \sprintf('%s($context["%s"] ?? null)', $line, $name);
     }
 
+    /**
+     * @deprecated since Twig 3.13
+     */
     protected function getAttributeGetter()
     {
+        trigger_deprecation('twig/twig', '3.13', 'Method "%s()" is deprecated.', __METHOD__);
```

Comment:
the deprecation message (and the `@deprecated` tag) should mention the replacement

Reply:
There is none. This method was only used by one subclass which does not really justify its presence on an abstract test case class.

Reply:
This is also used in other projects (like Symfony).

The whole point is to make it easier to write tests that support multiple Twig versions by making the NodeTestCase shipped in Twig itself responsible for knowing how a variable access gets compiled.
This was very important during the migration from Twig 2 to Twig 3 where the compilation output changed (Twig 3 to Twig 4 changes the compilation output for the strict variable getter, for which we don't have a similar method in the test case, but maybe we should add it to avoid making downstream projects add it, like in the TransNodeTest of Symfony)

Reply:
well, this one is about the `.` operator, not variable access, but the reason is the same (btw, without that method, the migration from `twig_get_attribute` to `CoreExtension::getAttribute` would have been painful for the ecosystem)

Reply:
All right, I'll add a static replacement.